src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1320,6 +1320,84 @@ static void Main()
         }
 
         [Fact]
+        [WorkItem(18554, "https://github.com/dotnet/roslyn/issues/18554")]
+        public void AssigningIntoIndexers()
+        {
+            string source = @"
+class C
+{
+    int field;
+    ref int this[int i, long opt = 1]
+    {
+        get
+        {
+            System.Console.Write(i); return ref field;
+        }
+    }
+
+    int this[long j, int opt = 1]
+    {
+        set
+        {
+            System.Console.Write(j + value.ToString());
+        }
+    }
+
+    static void Main()
+    {
+        var c = new C();
+        (c[i: 1], c[j: 2]) = c;
+    }
+
+    public void Deconstruct(out int a, out int b)
+    {
+        a = 1;
+        b = 2;
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "122", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(18554, "https://github.com/dotnet/roslyn/issues/18554")]
+        public void AssigningTupleIntoIndexers()
+        {
+            string source = @"
+class C
+{
+    int field;
+    ref int this[int i, long opt = 1]
+    {
+        get
+        {
+            System.Console.Write(i); return ref field;
+        }
+    }
+
+    int this[long j, int opt = 1]
+    {
+        set
+        {
+            System.Console.Write(j + value.ToString());
+        }
+    }
+
+    static void Main()
+    {
+        var c = new C();
+        (c[i: 1], c[j: 2]) = (1, 2);
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "122", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics();
+        }
```

Comment:
The `TransformIndexerAccess` shows that the subtler cases look like this: `Collection[z : Z(), x : X(), y : 123]`.
Could you add a test for this in a deconstruction, verifying the execution order (print to console whenever X, or Z are evaluated, which we expect to be just once, before the right-hand-side or `Deconstruct` would be invoked)?